src/audio/codec_adapter/codec/dcblock.c

```diff
@@ -0,0 +1,383 @@
+			ret = -EINVAL;
+			goto ret;
+		}
+		/* Obtain next parameter, it starts right after the preceding one */
+		data = (char *)data + param->size;
+		size -= param->size;
+	}
+
+	comp_dbg(dev, "apply_config() done");
+ret:
+	return ret;
+}
+
+int dcblock_codec_init(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct dcblock_codec_data *cd = NULL;
+
+	comp_info(dev, "dcblock_codec_init()");
+
+	cd = codec_allocate_memory(dev, sizeof(struct dcblock_codec_data), 0);
+	if (!cd) {
+		comp_err(dev, "dcblock_codec_init(): failed to allocate memory for dcblock codec data");
+		return -ENOMEM;
+	}
+
+	codec->private = cd;
+
+	memset(cd->state, 0, sizeof(cd->state));
+	set_passthrough(cd);
+	cd->dcblock_func = NULL;
+
+	comp_dbg(dev, "dcblock_codec_init() done");
+
+	return 0;
+}
+
+int dcblock_codec_prepare(struct comp_dev *dev)
+{
+	int ret = 0;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct dcblock_codec_data *cd = codec->private;
+	struct comp_buffer *sourceb;
+	enum sof_ipc_frame source_format;
+
+	comp_info(dev, "dcblock_codec_prepare()");
+
+	if (codec->state == CODEC_PREPARED)
+		goto done;
+
+	if (!codec->s_cfg.avail && !codec->s_cfg.size) {
+		comp_err(dev, "dcblock_codec_prepare() no setup configuration available!");
+		ret = -EIO;
+		goto err;
+	} else if (!codec->s_cfg.avail) {
+		comp_warn(dev, "dcblock_codec_prepare(): no new setup configuration available, using the old one");
+		codec->s_cfg.avail = true;
+	}
+	ret = apply_config(dev, CODEC_CFG_SETUP);
+	if (ret) {
+		comp_err(dev, "dcblock_codec_prepare() error %x: failed to applay setup config",
+			 ret);
+		goto err;
+	}
+	/* Do not reset nor free codec setup config "size" so we can use it
+	 * later on in case there is no new one upon reset.
+	 */
+	codec->s_cfg.avail = false;
+
+	/* DC Filter component will only ever have one source and sink buffer */
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+
+	/* get source data format and sample bytes */
+	source_format = sourceb->stream.frame_fmt;
+
+	cd->dcblock_func = dcblock_find_func(source_format);
+	if (!cd->dcblock_func) {
+		comp_err(dev, "dcblock_codec_prepare(): no processing function matching frames format");
+		ret = -EINVAL;
+		goto err;
+	}
+
+	comp_dbg(dev, "dcblock_codec_prepare() done");
+
+	return 0;
+err:
+done:
+	return ret;
+}
+
+int dcblock_codec_process(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct dcblock_codec_data *cd = codec->private;
+
+	comp_dbg(dev, "dcblock_codec_process()");
+
+	cd->dcblock_func(dev, codec->cpd.in_buff, codec->cpd.out_buff, codec->cpd.avail, &codec->cpd.produced);
+
+	comp_dbg(dev, "dcblock_codec_process() done");
+
+	return 0;
+}
+
+int dcblock_codec_apply_config(struct comp_dev *dev)
+{
+	return apply_config(dev, CODEC_CFG_RUNTIME);
+}
+
+int dcblock_codec_reset(struct comp_dev *dev)
+{
+	int ret;
+
+	codec_free_all_memory(dev);
```

Comment:
this doesn't seem right, why are freeing on reset? isn't that what free is for? I feel like memset is the proper thing here, otherwise we are constantly allocing and freeing memory on pipeline start and stop, sounds like a good way to fragment memory.